### PR TITLE
Buyers Guide featured articles can now include campaign pages

### PR DESCRIPTION
**Pull request by danielfmiranda**
# Description

Related PRs/issues: #10149 

This PR allows users to select `BuyersGuideCampaignPages` alongside `BuyersGuideArticlePages` as either a **main** or **supporting** feature on the PNI homepage's hero section.

image

Since the PNI homepage hero section is no longer specific to Article pages, any reference of the word "Article" is updated to "Page" in related fields/methods/templates.  ([discussed with design in original ticket](https://github.com/MozillaFoundation/foundation.mozilla.org/issues/10149#issuecomment-1442619240))

This PR also adds some tests for selecting campaign pages as both the main and supporting features. 

# Checklist

**Tests**
- [x] Is the code I'm adding covered by tests?
- [x] Did I update or add new fake data? (updated factories accordingly)
- [ ] Are my changes [backward-compatible](https://github.com/mozilla/foundation.mozilla.org/blob/main/docs/workflow.md#django-migrations-what-to-do-when-working-on-backward-incompatible-migrations). If not, did I schedule a deploy with the rest of the team?

### CODE REVIEW

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py
Comment (danielfmiranda) [inline network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py:57]: Using the [wagtail docs](https://docs.wagtail.org/en/stable/reference/pages/panels.html#pagechooserpanel) for reference, I updated the FK to point to wagtailcore.Page as this will allow us to select from more than one page type.

We then limit the choices to BuyersGuideArticlePage and BuyersGuideCampaignPage in this fields "PageChooserPanel" definition:

https://github.com/MozillaFoundation/foundation.mozilla.org/blob/4f75c5fd496d56e533157b38061ff08e07a03135/network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py#L102-L104

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py
Comment (danielfmiranda) [inline network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py:451]: Using the [wagtail docs](https://docs.wagtail.org/en/stable/reference/pages/panels.html#pagechooserpanel) for reference, I updated the FK to point to `wagtailcore.Page `as this will allow us to select from more than one page type. 

**Note:** Existing choices in this field remain even after running migrations

We then limit the choices to `BuyersGuideArticlePage` and `BuyersGuideCampaignPage` in this fields `PageChooserPanel` definition:

https://github.com/MozillaFoundation/foundation.mozilla.org/blob/4f75c5fd496d56e533157b38061ff08e07a03135/network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py#L458-L461

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py
Comment (danielfmiranda) [inline network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py:397]: Since we are now pointing to `wagtailcore.Page`, we use wagtail's [page.specific](https://docs.wagtail.org/en/stable/reference/pages/model_reference.html#wagtail.models.Page.specific) method to return the page in its subclassed form (`BuyersGuideCampaignPage`/`BuyersGuideArticlePage`) with all its field values for use in the template.

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py
Comment (danielfmiranda) [inline network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py:57]: Using the [wagtail docs](https://docs.wagtail.org/en/stable/reference/pages/panels.html#pagechooserpanel) as reference, I updated the FK relationship in **both fields** to point to `wagtailcore.Page`. This allows us to select from more than one page type.


The choices are then limited to `BuyersGuideCampaignPage` and `BuyersGuideArticlePage` in both fields `PageChooserPanel` definition:

https://github.com/MozillaFoundation/foundation.mozilla.org/blob/4f75c5fd496d56e533157b38061ff08e07a03135/network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py#L102-L105

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py
Comment (danielfmiranda) [inline network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py:397]: Since we are now pointing to `wagtailcore.Page`, we use wagtail's [page.specific](https://docs.wagtail.org/en/stable/reference/pages/model_reference.html#wagtail.models.Page.specific) method to return the page in its subclassed form (`BuyersGuideCampaignPage`/`BuyersGuideArticlePage`).

Not using .specific will return the page as a wagtailcore.Page model, which does not include all of the fields we need for use in the template.

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py
Comment (danielfmiranda) [inline network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py:395]: Hi @tbrlpld I tried to update the function annotations accordingly after reading about them, however, I was hoping that you could double check my updates. 

Thanks!

File: network-api/networkapi/templates/fragments/buyersguide/featured_page.html
Comment (danielfmiranda) [inline network-api/networkapi/templates/fragments/buyersguide/featured_page.html:2]: Double checked and can confirm that there is no JS referencing this ID 👍

**Comment by danielfmiranda** [comment]
Hi @tbrlpld just realized that my changes might be backward incompatible as they feature renaming of some fields, will bring this up during stand up!

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py
Comment (tbrlpld) [reply network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py:57]: Thanks for the explanation and the reference to the docs 👍

**Review by tbrlpld — COMMENTED** [review]
Thanks for this @danielfmiranda 

I have left a few questions and comments to your points. 

Good catch about the backward incompatibility. I wonder if we want to go the route of updating this to be backward compatible, which would mean data migrations for the orderable relation models will be needed. Or is we should disable pre-boot for the deployment and be ok with the small downtime 🤔 One to discuss with @mtdenton

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py:395]: Thanks for the suggestion Tibor! Updated 👍

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py:403]: Also updated

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py:410]: Thanks @tbrlpld, in the template, we are only rendering the title of these pages so removing the `.specific` still works. 
I have updated the code to not include the `.specific` method here as it is not needed

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py:70]: Updated

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py:66]: 👍 updated

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py:415]: Hi Tibor, the only field that we use from these pages are their titles, so there is no need for the `.specific` method. 

I have removed the `.specific` and just left the updated naming

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py:110]: Updated to Title Case to match the rest!

File: network-api/networkapi/wagtailpages/tests/buyersguide/test_homepage.py
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/tests/buyersguide/test_homepage.py:23]: Thanks for the catch, updated!

File: network-api/networkapi/wagtailpages/tests/buyersguide/test_homepage.py
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/tests/buyersguide/test_homepage.py:428]: Great idea! Updated

File: network-api/networkapi/wagtailpages/tests/buyersguide/test_homepage.py
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/tests/buyersguide/test_homepage.py:439]: Updated

File: network-api/networkapi/wagtailpages/tests/buyersguide/test_homepage.py
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/tests/buyersguide/test_homepage.py:462]: Updated this one as well as the campaign page equivalent 👍

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py:410]: Thanks Tibor! I have updated the comment.

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py:415]: Hi @tbrlpld, is this referencing the function `get_featured_articles`?

If so, I think that it is OK to leave for now. 

It has only been requested that we update fields in the "hero" section to be able to select from both page types. 
The `featured_articles` relationship should still only point to a `BuyersGuideArticlePage`, meaning that we do not need `.specific` here.

File: network-api/networkapi/wagtailpages/factory/buyersguide.py
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/factory/buyersguide.py:457]: Updated to add a campaign page to the supporting pages section!

File: network-api/networkapi/wagtailpages/factory/buyersguide.py
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/factory/buyersguide.py:89]: Hi Tibor, great question. At this moment, I cant think of any case where we should implement the selection here. 

I think that the Article page is a good default option, and if we ever need the supporting page to be a Campaign page we can pass it in as the `supporting_page` like I do in the tests and buyersguide factory file.

File: network-api/networkapi/templates/fragments/buyersguide/featured_page.html
Comment (danielfmiranda) [inline network-api/networkapi/templates/fragments/buyersguide/featured_page.html:2]: @tbrlpld  In response to your comment regarding the ID change from `featured-article` to `featured-pages`:

After speaking with design, I found that the ID is used for GTM to track link clicks in this DIV. 

I have the OK from design to change this to `featured-pages` as long as I update GTM to match when this gets merged.

Thanks!

**Comment by danielfmiranda** [comment]
Hi @tbrlpld thanks for taking a look at this PR!

I have made updates based on your feedback and re-requested review. I also left a couple of the conversations open where I had questions.

Thanks!

File: network-api/networkapi/templates/fragments/buyersguide/featured_page.html
Comment (tbrlpld) [reply network-api/networkapi/templates/fragments/buyersguide/featured_page.html:2]: Thanks for checking @danielfmiranda. Since you already need to change it in GTM to, I would suggest to also move to the more obvious `data-gtm` approach (which makes clear why it exists). We might want to add a the `data-gtm` approach to the docs somewhere 🤔

Comment (danielfmiranda) [reply network-api/networkapi/templates/fragments/buyersguide/featured_page.html]: Sounds good! I have updated it to now use `data-gtm` instead of `id`. 

Regarding updating the docs, I will create a ticket to update them!

**Comment by danielfmiranda** [comment]
Hi @tbrlpld @mtdenton!

In order to make sure this change is backwards compatible, I have split the front-end and back-end changes into two PRs, as suggested in our [backwards-compatibility workflow](https://github.com/MozillaFoundation/foundation.mozilla.org/blob/main/docs/workflow.md#django-migrations-what-to-do-when-working-on-backward-incompatible-migrations):


**Back-end changes:** https://github.com/MozillaFoundation/foundation.mozilla.org/pull/10245
**Front-end changes:** https://github.com/MozillaFoundation/foundation.mozilla.org/pull/10246

File: network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py
Comment (tbrlpld) [reply network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py:420]: Fair.

**Comment by tbrlpld** [comment]
> In order to make sure this change is backwards compatible, I have split the front-end and back-end changes into two PRs, as suggested in our [backwards-compatibility workflow](https://github.com/MozillaFoundation/foundation.mozilla.org/blob/main/docs/workflow.md#django-migrations-what-to-do-when-working-on-backward-incompatible-migrations):

@danielfmiranda I am not your about the splitting between frontend and backend. The split really needs to be different steps and they should be 3 PRs. 

1. Adding a new field with the desired name and copy all data from the existing field into the new one (we still keep the old field around). 
2. Switch all  references of the old field to use the new field (this might involve changes on the frontend, but also backend).
3. Delete the old field. 

But, as discussed on Slack, I would also be ok with disabling preboot for the deployment and relying on the frontend cache to cover the downtime.

**Comment by danielfmiranda** [comment]
Hi Tibor thanks for your feedback about combining the migrations into one file, and also about using the specific method on the queryset!

I updated this PR with your feedback about the migrations.

However, I was unable to get the specific method on the query to work. I believe its because the queryset is populated with SupportingPageRelations and not with Page objects.

**Review by tbrlpld — APPROVED** [review]

**Comment by danielfmiranda** [comment]
[NOTE]: Please don't merge!

Interestingly, the PR checks were frozen for a majority of Friday. When they finally unblocked and allowed me to merge, I found that I do not have access to disable preboot, as mentioned in comments above. 

Going to bring this up on Monday.

Thanks!

**Comment by tbrlpld** [comment]
@danielfmiranda Have you been able to disable preboot on staging?

**Pull request merged**
